sources/web/datalab/polymer/components/datalab-terminal/datalab-terminal.ts

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2017 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+/**
+ * Type declaration for xterm.js
+ */
+declare class Terminal {
+  constructor(parameters?: object)
+  open(element: HTMLElement, focus: boolean): void;
+  write(data: string): void;
+  clear(): void;
+  reset(): void;
+  resize(cols: number, rows: number): void;
+  on(event: string, handler: Function): void;
+}
+
+/**
+ * Terminal element for Datalab.
+ * This element uses xterm.js for the front end terminal element, and opens a websocket
+ * connection to a Jupyter terminal on the backend. Although Jupyter supports multiple
+ * terminals, this element keeps an instance of exactly one and uses it to make the user's
+ * experience simpler.
+ * If the user closes the terminal session, either by typing 'exit' or ctrl+d, this
+ * element will automatically reset the terminal.
+ */
+class TerminalElement extends Polymer.Element {
+
+  private _xterm: Terminal;
+  private _wsConnection: WebSocket;
+  private _boundResizeHandler: EventListenerObject;
+  private _charHeight: number;
+  private _charWidth: number;
+
+  static get is() { return "datalab-terminal"; }
+
+  ready() {
+    super.ready();
+    const self = this;
+
+    // Will be called after the custom element is done rendering.
+    window.addEventListener('WebComponentsReady', function() {
+      // Use the size helper element to get the height and width of a character. This
+      // makes changing the style simpler, instead of hard-coding these values.
+      self._charHeight = self.$.sizeHelper.clientHeight;
+      self._charWidth = self.$.sizeHelper.clientWidth / 10; // The element has 10 characters.
+
+      self._boundResizeHandler = self._resizeHandler.bind(self);
```

Comment:
This is here so I can call `removeEventListener`. The binding, regardless whether it's done explicitly or implicitly using an arrow function, generates a new function reference, which is what needs to be passed in the remove call.